### ISSUE TITLE
Future of ecmaFeatures.globalReturn

### ISSUE THREAD

**Issue by nzakas** [eslint/espree#525]
**Problem:** `globalReturn` can't be used with `sourceType:module`, and right now in ESLint we will detect this combination and set `globalReturn` to `false`. Arguably, this behavior is undesirable because ESLint is silently fixing a problem rather than letting the user know that it happened.

We recently just implemented `sourceType: commonjs`, which effectively makes `ecmaFeatures.globalReturn` obsolete, so there are several options we could pursue going forward:

1. We could remove `ecmaFeatures.globalReturn` altogether. I think this is the cleanest solution now that we have `sourceType: commonjs`, however, we would probably want to throw an error if `ecmaFeatures.globalReturn` is specified to let people know that it has been removed.
2. We could keep `ecmaFeatures.globalReturn` but throw an error if it is used with `sourceType: module`. This error would bubble up to ESLint and to the user, allowing them to fix their configuration.

In either case, we could remove the logic from ESLint completely.

Thoughts?

**Comment by btmills** [comment]
Option 1 seems cleanest to me as well. We'd be able to give a very clear and actionable error message: `ecmaFeatures.globalReturn` is gone, so you should set `sourceType: "commonjs"` instead.

**Comment by mdjermanovic** [comment]
If we are not aware of any use cases for `globalReturn = true` other than commonjs modules, then option 1 seems best to me.

**Comment by nzakas** [comment]
Marking as accepted.

**Comment by nzakas** [comment]
We aren’t quite ready to implement this yet. We will need to wait for the new config system to roll out and give advance notice before we remove any existing functionality.

**Comment by amareshsm** [comment]
Is it good to go now?

**Comment by nzakas** [comment]
Not yet. This will need to happen in the v10.0.0 timeframe, once the old config system has been removed.

**Comment by sam3k** [comment]
TSC group has confirmed this for `v10`

**Comment by Pixel998** [comment]
I'll work on this.

**Comment by mdjermanovic** [comment]
Shall we also remove `nodejsScope` option of `eslint-scope`? It's redundant with `sourceType: "commonjs"`.

I'm including this here because it's related to the use of `globalReturn` in ESLint, which is going to be removed as part of this task. So, ESLint will no longer use `nodejsScope`, and there's an alternative for other consumers of eslint-scope.

**Comment by Pixel998** [comment]
I completely agree with @mdjermanovic. I'm happy to include this change in #692 or open a separate PR—whichever is preferred.

**Comment by mdjermanovic** [comment]
I opened a separate issue about `nodejsScope`: https://github.com/eslint/js/issues/697

**Comment by erosman** [comment]
Please note that Greasemonkey, Tampermonkey, Violentmonkey, and FireMonkey (next release v3.0) (as well as Node.js), wrap userscripts in an IIFE which avoids JavaScript syntax errors such as `SyntaxError: return not in function`.

As also mentioned in https://github.com/eslint/eslint/issues/19881

- non-strict mode requires `sourceType: "script"`
- Top-level `await` requires `sourceType: "module"`
- `globalReturn` requires `sourceType: "commonjs"`

Removing `ecmaFeatures.globalReturn` would further complicate ESLint implementation in the aforementioned environments.

**Comment by mdjermanovic** [comment]
> Please note that Greasemonkey, Tampermonkey, Violentmonkey, and FireMonkey (next release v3.0) (as well as Node.js), wrap userscripts in an IIFE which avoids JavaScript syntax errors such as `SyntaxError: return not in function`.

I think we could consider supporting these environments at some point, but still proceed with removing the `globalReturn` option because it just allows syntax (`return` at the top level) without providing execution context. I believe it would be better to provide the context (e.g., "this code is the body of a function that is executed in a script" or  "this code is the body of an async function that is executed in a script") in another option or perhaps as new values for `sourceType`.

**Comment by mdjermanovic** [comment]
In the 2025-10-16 TSC meeting, we decided to postpone this change to ESLint v11.

Before making this change, we should discuss and implement support for the use case described in https://github.com/eslint/js/issues/525#issuecomment-3405494567.

**Comment by erosman** [comment]
A few ideas to consider ...

1. a new `sourceType: "userscript"`
2. `sourceType` independent `parserOptions` e.g.
   - globalReturn: true,
   - strict: true,
   - topLevelAwait: true,
3. `sourceType` independent `rules` options e.g.
   - global-return: ["error", "warn", "off"],
   - strict: ["error", "warn", "off"],
   - top-level-await: ["error", "warn", "off"],

The `rules` options might prove to be the most versatile.

**Comment by nzakas** [comment]
@erosman These aren't rules, though, they are parsing modes, which happen at a different level of linter. It's not just about warning for syntax or not, it's about changing how scopes are evaluated completely. 

`sourceType: userscript` is the cleanest option, but that would mean assuming all userscript extensions act the same way and will continue to do so.

Independent parser options are a problem, as already mentioned, because it's difficult to reconcile them with `sourceType: module`.

So, this is something that will need a great deal of thought to move forward.

**Comment by erosman** [comment]
> These aren't rules, though, they are parsing modes, which happen at a different level of linter.

I understand... I thought that it might be possible to process normally but ignore or output the error based on the rule.


> `sourceType: userscript` is the cleanest option, but that would mean assuming all userscript extensions act the same way and will continue to do so.

I agree. Besides the fact that too many `sourceType` variants are not desirable (and more work to maintain), it may create more issues than it solves.

**Comment by mdjermanovic** [comment]
A comprehensive solution that doesn't involve configuring individual language features could be a new option that would indicate that the given code is implicitly wrapped in a function, and which kind of function it is (sync/async).

High-level overview (there would be more details to discuss):

1. `{ sourceType: "script", wrapperFunction: "sync" }`. Scopes: Global > Function. Enabled syntax: `return` at the top level.
2. `{ sourceType: "script", wrapperFunction: "async" }`. Same as 1., plus `await` at the top level.
3. `{ sourceType: "module", wrapperFunction: "sync" }`. Scopes: Global > Module > Function. Enabled syntax: `return` at the top level, `import.meta`; not `import` and `export` declarations.
4. `{ sourceType: "module", wrapperFunction: "async" }`. Same as 3., plus `await` at the top level.

**Comment by erosman** [comment]
@mdjermanovic Sounds good.

If I understand correctly, `wrapperFunction` would have "3" states:

- `wrapperFunction: null` (default, no wrapper processing)
- `wrapperFunction: "sync"`
- `wrapperFunction: "async"`

**Comment by mdjermanovic** [comment]
Yes, the default would be no wrapper processing. That is when the `wrapperFunction` property is omitted or set to `undefined` (maybe `null` as well; we could discuss that when it comes to details, if the team is in favor of this solution).
